core/multivm_processor.go

```diff
@@ -0,0 +1,163 @@
+			break Loop
+		case sputnikvm.RequireAccount:
+			address := ret.Address()
+			if statedb.Exist(address) {
+				vm.CommitAccount(address, new(big.Int).SetUint64(statedb.GetNonce(address)),
+					statedb.GetBalance(address), statedb.GetCode(address))
+			} else {
+				vm.CommitNonexist(address)
+			}
+		case sputnikvm.RequireAccountCode:
+			address := ret.Address()
+			if statedb.Exist(address) {
+				vm.CommitAccountCode(address, statedb.GetCode(address))
+			} else {
+				vm.CommitNonexist(address)
+			}
+		case sputnikvm.RequireAccountStorage:
+			address := ret.Address()
+			key := common.BigToHash(ret.StorageKey())
+			if statedb.Exist(address) {
+				value := statedb.GetState(address, key).Big()
+				key := ret.StorageKey()
+				vm.CommitAccountStorage(address, key, value)
+			} else {
+				vm.CommitNonexist(address)
+			}
+		case sputnikvm.RequireBlockhash:
+			number := ret.BlockNumber()
+			hash := bc.GetBlockByNumber(number.Uint64()).Hash()
+			vm.CommitBlockhash(number, hash)
+		}
+	}
+
+	// VM execution is finished at this point. We apply changes to the statedb.
+
+	for _, account := range vm.AccountChanges() {
+		switch account.Typ() {
+		case sputnikvm.AccountChangeIncreaseBalance:
+			address := account.Address()
+			amount := account.ChangedAmount()
+			statedb.AddBalance(address, amount)
+		case sputnikvm.AccountChangeDecreaseBalance:
+			address := account.Address()
+			amount := account.ChangedAmount()
+			balance := new(big.Int).Sub(statedb.GetBalance(address), amount)
+			statedb.SetBalance(address, balance)
+		case sputnikvm.AccountChangeRemoved:
+			address := account.Address()
+			statedb.Suicide(address)
+		case sputnikvm.AccountChangeFull:
+			address := account.Address()
+			code := account.Code()
+			nonce := account.Nonce()
+			balance := account.Balance()
+			statedb.SetBalance(address, balance)
+			statedb.SetNonce(address, nonce.Uint64())
+			statedb.SetCode(address, code)
+			for _, item := range account.ChangedStorage() {
+				statedb.SetState(address, common.BigToHash(item.Key), common.BigToHash(item.Value))
+			}
+		case sputnikvm.AccountChangeCreate:
+			address := account.Address()
+			code := account.Code()
+			nonce := account.Nonce()
+			balance := account.Balance()
+			statedb.SetBalance(address, balance)
+			statedb.SetNonce(address, nonce.Uint64())
+			statedb.SetCode(address, code)
+			for _, item := range account.Storage() {
+				statedb.SetState(address, common.BigToHash(item.Key), common.BigToHash(item.Value))
+			}
+		default:
+			panic("unreachable")
+		}
+	}
+	for _, log := range vm.Logs() {
+		statelog := evm.NewLog(log.Address, log.Topics, log.Data, header.Number.Uint64())
+		statedb.AddLog(statelog)
+	}
+	usedGas := vm.UsedGas()
+	totalUsedGas.Add(totalUsedGas, usedGas)
+
+	receipt := types.NewReceipt(statedb.IntermediateRoot().Bytes(), totalUsedGas)
+	receipt.TxHash = tx.Hash()
+	receipt.GasUsed = new(big.Int).Set(totalUsedGas)
+	if MessageCreatesContract(tx) {
+		from, _ := tx.From()
```

Comment:
Don't we already have `from` from up top on line 23?

Reply:
Does golang has variable scope? So this line defines a new `from` variable, and out of this `if` scope `from` would still be the above one, right? It is really common to do this kind of thing in Rust, does it work in go?

Anyway, `from` is never used below again so this shouldn't cause any trouble.

Reply:
I feel like I might be not understanding something here

Reply:
Ooops you're right. Fixed by removing this line.